File: strategy/generator/signal/flip/supertrend.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.smooth import SmoothATR
from core.models.source import SourceType
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class SupertrendFlipSignal(Signal):
    type: SignalType = SignalType.SupFlip
    source_type: Parameter = StaticParameter(SourceType.HL2)
    smooth_atr: Parameter = StaticParameter(SmoothATR.SMMA)
    period_atr: Parameter = StaticParameter(8.0)
    factor: Parameter = StaticParameter(0.86)

File: ta_lib/strategies/signal/src/flip/supertrend.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use trend::supertrend;

pub struct SupertrendFlipSignal {
    source: SourceType,
    smooth_atr: Smooth,
    period_atr: usize,
    factor: f32,
}

impl SupertrendFlipSignal {
    pub fn new(source: SourceType, smooth_atr: Smooth, period_atr: f32, factor: f32) -> Self {
        Self {
            source,
            smooth_atr,
            period_atr: period_atr as usize,
            factor,
        }
    }
}

impl Signal for SupertrendFlipSignal {
    fn lookback(&self) -> usize {
        self.period_atr
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let (direction, _) = supertrend(
            &data.source(self.source),
            data.close(),
            &data.atr(self.smooth_atr, self.period_atr),
            self.factor,
        );

        (direction.cross_over(&ZERO), direction.cross_under(&ZERO))
    }
}
